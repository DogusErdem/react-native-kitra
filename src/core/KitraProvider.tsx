import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
