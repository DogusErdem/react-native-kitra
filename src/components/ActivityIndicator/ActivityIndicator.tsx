import { StyleSheet, View, ActivityIndicator as RNActivityIndicator } from 'react-native';
import type { ActivityIndicatorProps, FCCWD } from '../../types';
import { opacity } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';

const styles = StyleSheet.create({
  centeredView: {
    zIndex: 100,
    flex: 1,
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  modalView: {
    borderRadius: 20,
    width: 150,
    height: 150,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
const ActivityIndicator:FCCWD<ActivityIndicatorProps> = ({ theme, children }) => (
  <View style={[styles.centeredView, { backgroundColor: opacity(String(theme?.lightBlack), 80) }]}>
    <View style={[styles.modalView]}>
      {children || <RNActivityIndicator size="large" color={theme?.primary} /> }
    </View>
  </View>
);

export default applyDefaults(ActivityIndicator);
