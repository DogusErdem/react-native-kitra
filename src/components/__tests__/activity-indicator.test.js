import { fireEvent, render, screen } from '@testing-library/react-native';
import { Text } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import ActivityIndicator from '../ActivityIndicator/ActivityIndicator';
/* eslint-disable no-undef */

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('ActiviyIndicator', () => {
  it('renders avtivity indicator with children', () => {
    // eslint-disable-next-line react/no-children-prop
    const tree = renderWithContext(<ActivityIndicator children={<Text>TEST</Text>} />);
    expect(tree).toMatchSnapshot();
  });
  it('renders avtivity indicator', () => {
    const tree = renderWithContext(<ActivityIndicator />);
    expect(tree).toMatchSnapshot();
  });
});
