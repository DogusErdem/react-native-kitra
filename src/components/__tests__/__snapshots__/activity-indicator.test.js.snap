// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiviyIndicator renders avtivity indicator 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
          "zIndex": 100,
        },
        {
          "backgroundColor": "rgba(64,64,64,0.8)",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderRadius": 20,
            "height": 150,
            "width": 150,
          },
        ]
      }
    >
      <ActivityIndicator
        color="#8973CD"
        size="large"
      />
    </View>
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`ActiviyIndicator renders avtivity indicator with children 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
          "zIndex": 100,
        },
        {
          "backgroundColor": "rgba(64,64,64,0.8)",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderRadius": 20,
            "height": 150,
            "width": 150,
          },
        ]
      }
    >
      <Text>
        TEST
      </Text>
    </View>
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
